// kaspa-rpc-client/dist/lib/ClientWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientWrapper = void 0;
const Client_1 = require("./Client");
class ClientWrapper {
    constructor({ hosts, verbose }) {
        this.clients = hosts.map((host) => new Client_1.Client({ host, verbose }));
    }
    async initialize() {
        await Promise.any(this.clients.map((client) => client.connect()));
        await Promise.any(this.clients.map((client) => client.ping()));
    }
    disconnect() {
        this.clients.forEach((client) => client.disconnect());
    }
    async sleep(ms = 1000) {
        await new Promise((resolve) => setTimeout(resolve, ms));
    }
    async getClient() {
        let client = this.clients.find((client) => client.isReady());
        while (!client) {
            await this.sleep();
            await this.initialize();
            client = this.clients.find((client) => client.isReady());
        }
        return client;
    }
}
exports.ClientWrapper = ClientWrapper;
